### ISSUE TITLE
Lack of important features

### ISSUE THREAD

**Issue by polygamma** [polygamma/aurman#25]
I am currently quite satisfied with aurman's development status, I can't think of any *important* missing features.

Can anyone else think of something I haven't thought of that really should be supported?

**Comment by rmarquis** [comment]
Powerful search. Cower/auracle regex feature is currently unequaled, and this is the single reason I can't do without these.

**Comment by polygamma** [comment]
I bet that's not too hard to implement with python, since it's python ;)
Thanks! I think I would not use that, but since people are actually using regex and pacman itself is supporting regex, aurman should support that, too. So: Going to implement!

**Comment by rmarquis** [comment]
Yes, python is just the right tool for this feature.

Also, take note that the AUR has a limitation of minimum 2 characters when sending request, so you'll need to take this into account (see cower code).

>  I think I would not use that

I'm a weird guy. I heavily dislike the popular yaourt-like search list feature that many helpers seem to implement, and a high percentage of my search involve regex. I might actually not be the right person to ask feature requests :)

**Comment by ChuckDaniels87** [comment]
Just two come to my mind right now:
* ZSH Completion. :smiley:
* `aurman` cache clean-up using a command-line argument. (maybe two non-installed aur packages and full clean-up)

Great work @polygamma!

**Comment by polygamma** [comment]
See: https://github.com/polygamma/aurman/commit/07a7caf1c69dc1d466ccddff58e5b47d8fb14738

**Comment by polygamma** [comment]
@ChuckDaniels87, I have looked into zsh completion, but that seems like hardcoding very much stuff... e.g. pacaur zsh completion is about ~600 lines of code. Compare that to my nice python bash-completion...

Maybe there is a nicer way of doing it, but currently I am unaware of it. I am going to look into it, for sure, but currently it's very disappointing to think about implementing that.

**Comment by ChuckDaniels87** [comment]
I will take a look to ZSH completion but I can't promise anything. :wink:

**Comment by polygamma** [comment]
@ChuckDaniels87 , see: https://github.com/polygamma/aurman/commit/4da16fad339ec330ef2cabaec77fdafb2b6a0e53

**Comment by ChuckDaniels87** [comment]
`aurman -Sc --aur` removed everything in `$HOME/.cache/aurman`, even currently installed packages.

**Comment by polygamma** [comment]
Cannot reproduce, works as expected here.

**Comment by AladW** [comment]
>and a high percentage of my search involve regex.

I guess you mean regex for `name-desc`? Regex support for `name` searches is trivial since you can just use `packages.gz`.

**Comment by polygamma** [comment]
@ChuckDaniels87 was able to reproduce it: https://github.com/polygamma/aurman/commit/915ccdc01616e75f09cf518c0c256918e810c8b9

**Comment by craftyguy** [comment]
I would like an option to always show the PKGBUILD for a package and not prompt. Or, at the very least, change the prompt to default to 'yes' and not 'no'. Reviewing PKGBUILDs is absolutely necessary, IMHO, since anyone can upload anything to AUR, and it's not a good idea to make it harder for folks to do that by making them perform an additional step.

**Comment by craftyguy** [comment]
Yea that would work for me, since I can just alias `aurman=aurman --edit`

**Comment by polygamma** [comment]
Okay, I see your point. I am going to implement that, but one more question: Just showing the differences by default, or also open the editor to edit the file afterwards?

**Comment by craftyguy** [comment]
If just showing differences by default, then an option called `--edit` doesn't make much sense. 

What would you think about changing the default behavior of aurman to always show differences (unless `--noedit`), then prompt after showing the differences to edit. 

This would not require the addition of a new command, but you could add an `--edit` that opens the PKGBUILD in the editor instead of showing the differences (I don't feel strongly about having this though).

Basically, I don't want to go through extra steps to see PKGBUILD diffs (or in the case of a new install, to see the full PKGBUILD + .install files) when updating apps from AUR since it makes it too easy to ignore potentially scary stuff in PKGBUILDs :smile:

**Comment by polygamma** [comment]
See: https://github.com/polygamma/aurman/commit/2816197a68e0c0faa4e589b9fab669055f7f573a

I named the new flag `--show_changes` instead of `--edit`, since that is what it does.
You will be shown the changes without being asked, that's it.

**Comment by craftyguy** [comment]
Just tested and it works great. Thanks!

**Comment by polygamma** [comment]
You're very welcome :)

**Comment by AladW** [comment]
What if some other relevant file changes that's not the PKGBUILD or .install (e.g., a `.patch` file)?

**Comment by polygamma** [comment]
Then that changes are being shown

**Comment by polygamma** [comment]
Sorry, I read that wrong. That will not be shown, as it was not shown before

**Comment by AladW** [comment]
If it's included in the AUR repository it should be worth being shown, apart from SRCINFO perhaps which has no effect on anything client side. 

Additional prompts for different files when you already 3 million is really annoying though. Ideally you could show all files in a single window + header à la `git show` but for every object in the repository. I'm sure git has some native way to do this...

**Comment by polygamma** [comment]
@AladW see: https://github.com/polygamma/aurman/commit/55607312e30e2e2ce9398dd204b4a4a039eb2c8b

**Comment by craftyguy** [comment]
Could you omit .SRCINFO? That has no impact on the client side and showing it just creates noise.

**Comment by polygamma** [comment]
@craftyguy : https://github.com/polygamma/aurman/commit/cdb30237882809d0b8d46a9a74077be2c6e2b0e5

**Comment by craftyguy** [comment]
well, not really. if I run this with the shiny new `--show_changes` you added, then I would see the .SRCINFO (which, again, is useless to display to users)

**Comment by polygamma** [comment]
Are you definitely using the latest master branch version?

**Comment by polygamma** [comment]
Since this blew up quite a bit:

@rmarquis - regarding the regex search, have you tried it? And if so, working as expected?
@ChuckDaniels87 - has the bug with the deletion of folders of still installed packages during `aurman -Sc --aur` been fixed for you?

**Comment by ChuckDaniels87** [comment]
@polygamma  `aurman -Sc --aur` is working as expected now. Thanks!

**Comment by AladW** [comment]
So do I get it right that besides the three prompts for each package, when confirming to "any" you get... a prompt for every file? o_O

Besides that's imo a rather excessive amount of prompts, this also means there's no state during editing, i.e. you can't go back to previous edited files if you forgot something. Any modern console editor should support buffers and editing of multiple files in a single session, at the least.

**Comment by polygamma** [comment]
Well, how would you do it then? I, personally, am not a fan of doing that via console. If I see a PKGBUILD or whatever, which I want to change, I CTRL+C, edit those files and start again, hence I do not really care about that...

**Comment by AladW** [comment]
So say you have ros-lunar-desktop (I don't think I'm being unfair with such extreme examples since they are explicitely supported by `aurman`) and you want to change PKGBUILD 249 of 250, you bail out, edit, then restart with `--noedit` or go through 249 prompts again? IDK.

Note that I don't insist on you using some different mechanism, I just try to find the appeal behind all this promptage. I use `vifm`, though that has other drawbacks, such as it being [literally the only file manager](https://github.com/AladW/aurutils/issues/51#issuecomment-368171103) that is suitable for these purposes.

**Comment by polygamma** [comment]
Well, that is actually not a problem, since directly after answering "yes" to the question "are you fine with using .. ?" that is being saved. hence finding a problem in 249/250 is great, aurman would only show 249 and 250 again

**Comment by AladW** [comment]
Hm, I see. So it saves it to some configuration file (is that documented somewhere?). Even then, you still have the initial 700+ prompts but I guess it's better than redoing them all the time.

**Comment by polygamma** [comment]
For every package the SHA1 hash of the last seen and approved commit is being saved. see: https://github.com/polygamma/aurman/blob/master/src/aurman/classes.py#L889

And the fetching of the files: https://github.com/polygamma/aurman/blob/master/src/aurman/classes.py#L814

**Comment by polygamma** [comment]
@AladW - see: https://github.com/polygamma/aurman/commit/5d625e8b5aeb403ae981051bcfa10301b37b22ee

![](https://user-images.githubusercontent.com/20651500/37254228-83eee2e2-253b-11e8-89b1-0910717a8da8.png)

You may choose by yourself, which files you want to edit. Hence you are able to edit files twice etc.

**Comment by rmarquis** [comment]
@polygamma Sorry, I haven't had time to have a closer look yet.

**Comment by leosanvieira** [comment]
After reading the docs, i'm still not sure how i'm supposed to check updates for aur packages, i used to issue <code>pacaur -Qqum --devel --needed</code> with pacaur, did i miss something or is it not implemented yet?

**Comment by polygamma** [comment]
> All pacman operations are supported, but calling aurman with an operation besides --sync or -S will just pass the arguments to pacman.

So: `aurman` does not support `--devel` and `--needed` for `-Q`, and tbh I do not really see a reason for that. What is wrong with `aurman -Syu --aur --devel --needed` ? It will show the packages to be updated, you can just exit with entering **n** if you do not want to install those.

**Comment by leosanvieira** [comment]
Nothing wrong with <code>aurman -Su --aur --devel --needed</code>, my question is a just a matter of workflow as i use an extension that checks updates from time to time from both, official repos and aur, and lists the updates available. Pacaur has -Qqum and aurutils has aurcheck to do that, that's why i asked. After reading your answer, i understand you're not interested in this workflow and that's ok as it's your tool, after all.

**Comment by polygamma** [comment]
@leosanvieira You may use `aurmansolver` for your purpose. see: [aurman as dep solver](https://github.com/polygamma/aurman/wiki/Using-aurman-as-dependency-solver)

you would simply call `aurmansolver -Su --devel --aur --needed`
but update aurman to the latest version first, since I included `--devel` a few minutes ago.

when using `--devel` with the solver, it will not fetch new pkgbuilds, and thus only use what`s already in the cache folder of aurman, but that is exactly what you want, i guess?

the output of the solver is json, as being described in the wiki. you may just strip the info you do not need and hence put the output in a form you want it to be.

**Comment by polygamma** [comment]
I am closing this issue since it really blew up and discusses many things at the same time. The features being discussed here, are fine right now. For new requests, just open a new issue.

**Comment by AladW** [comment]
Thanks, https://github.com/polygamma/aurman/issues/25#issuecomment-372117135 is definitely an improvement.

**Comment by ChuckDaniels87** [comment]
Uhm, I've taken a look at Bash and ZSH completion. However, aurman Bash completion is not working on my system for some reason. What is it supposed to complete? I've tried the following commands and returns nothing:

`aurman -<TAB>`
`aurman -S<TAB>`
`aurman -Ss <TAB>`
`aurman -Q<TAB>`

Not sure if it is a problem with my system or with completion file.

**Comment by polygamma** [comment]
Dumb question: Have you `bash-completion` installed on your system? e.g.

```
[jonny@jonny-arch-pc ~]$ aurman -<TAB>
-D          --files     -Q          --remove    -U          --version
--database  -h          --query     -S          --upgrade   
-F          --help      -R          --sync      -V
```

**Comment by ChuckDaniels87** [comment]
I deleted that package in the past for some stupid reason that I can't recall right now. I'll try to make Zsh completion based on bash completion, and sorry for the noise!

**Comment by ur-n1ghtmar3** [comment]
Hello! I too have a problem with bash completion
It is installed in my system
```
[vlad@ArchVM ~]$ pacman -Q bash-completion
bash-completion 2.8-1
```
But when i try 
```
[vlad@ArchVM ~]$ aurman -<TAB>
```
I get
```
[vlad@ArchVM ~]$ aurman -2018-04-25 22:12:55,743 - parse_args - parse_pacman_args - ERROR -  is too short
2018-04-25 22:12:56,164 - parse_args - parse_pacman_args - ERROR -  is too short
```
I also can't autocomplete package names

**Comment by polygamma** [comment]
This is not the right place for your issue, I wont be answering here

**Comment by polygamma** [comment]
I do not understand what you are saying

**Comment by iboyperson** [comment]
A nice feature would be displaying of aur packages installed that have been flagged as out of date. This is not a super important feature but none the less it would be great to have.

**Comment by polygamma** [comment]
that feature is already implemented for "-Ss"

**Comment by polygamma** [comment]
also please stop using this issue for feature requests. it is closed for a reason
